File: wcl_dotnet/src/Wcl/Serde/WclSerializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Reflection;
using System.Text;
using Wcl.Core;
using Wcl.Eval;

namespace Wcl.Serde
{
    public static class WclSerializer
    {
        public static string Serialize(object value, bool pretty = false)
        {
            var sb = new StringBuilder();
            SerializeObject(value, sb, pretty, 0);
            return sb.ToString();
        }

        private static void SerializeObject(object? value, StringBuilder sb, bool pretty, int indent)
        {
            if (value == null) { sb.Append("null"); return; }

            var type = value.GetType();

            if (type == typeof(string)) { sb.Append('"').Append(EscapeString((string)value)).Append('"'); return; }
            if (type == typeof(int) || type == typeof(long) || type == typeof(Int64))
            { sb.Append(value); return; }
            if (type == typeof(double)) { sb.Append(((double)value).ToString(CultureInfo.InvariantCulture)); return; }
            if (type == typeof(float)) { sb.Append(((float)value).ToString(CultureInfo.InvariantCulture)); return; }
            if (type == typeof(bool)) { sb.Append((bool)value ? "true" : "false"); return; }

            // List/Array
            if (value is IList list)
            {
                sb.Append('[');
                for (int i = 0; i < list.Count; i++)
                {
                    if (i > 0) sb.Append(", ");
                    SerializeObject(list[i], sb, pretty, indent);
                }
                sb.Append(']');
                return;
            }

            // Dictionary
            if (value is IDictionary dict)
            {
                if (pretty) SerializeMapPretty(dict, sb, indent);
                else SerializeMapCompact(dict, sb);
                return;
            }

            // POCO
            var props = type.GetProperties(BindingFlags.Public | BindingFlags.Instance);
            if (pretty)
            {
                var ind = new string(' ', indent);
                var inner = new string(' ', indent + 4);
                sb.AppendLine();
                foreach (var prop in props)
                {
                    if (!prop.CanRead) continue;
                    var val = prop.GetValue(value);
                    sb.Append(inner).Append(ToSnakeCase(prop.Name)).Append(" = ");
                    SerializeObject(val, sb, pretty, indent + 4);
                    sb.AppendLine();
                }
            }
            else
            {
                bool first = true;
                foreach (var prop in props)
                {
                    if (!prop.CanRead) continue;
                    if (!first) sb.AppendLine();
                    first = false;
                    var val = prop.GetValue(value);
                    sb.Append(ToSnakeCase(prop.Name)).Append(" = ");
                    SerializeObject(val, sb, false, 0);
                }
            }
        }

        private static void SerializeMapCompact(IDictionary dict, StringBuilder sb)
        {
            sb.Append('{');
            bool first = true;
            foreach (DictionaryEntry entry in dict)
            {
                if (!first) sb.Append(", ");
                first = false;
                sb.Append(entry.Key).Append(" = ");
                SerializeObject(entry.Value, sb, false, 0);
            }
            sb.Append('}');
        }

        private static void SerializeMapPretty(IDictionary dict, StringBuilder sb, int indent)
        {
            var inner = new string(' ', indent + 4);
            sb.AppendLine("{");
            foreach (DictionaryEntry entry in dict)
            {
                sb.Append(inner).Append(entry.Key).Append(" = ");
                SerializeObject(entry.Value, sb, true, indent + 4);
                sb.AppendLine();
            }
            sb.Append(new string(' ', indent)).Append('}');
        }

        private static string EscapeString(string s)
        {
            return s.Replace("\\", "\\\\").Replace("\"", "\\\"")
                    .Replace("\n", "\\n").Replace("\r", "\\r").Replace("\t", "\\t");
        }

        private static string ToSnakeCase(string name)
        {
            var sb = new StringBuilder();
            for (int i = 0; i < name.Length; i++)
            {
                if (char.IsUpper(name[i]))
                {
                    if (i > 0) sb.Append('_');
                    sb.Append(char.ToLower(name[i]));
                }
                else sb.Append(name[i]);
            }
            return sb.ToString();
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/serde/WclSerializer.java
package io.github.wiltaylor.wcl.serde;

import java.lang.reflect.Modifier;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class WclSerializer {
    private WclSerializer() {}

    public static String serialize(Object value, boolean pretty) {
        var sb = new StringBuilder();
        serializeObject(value, sb, pretty, 0);
        return sb.toString();
    }

    private static void serializeObject(Object value, StringBuilder sb, boolean pretty, int indent) {
        if (value == null) { sb.append("null"); return; }

        var type = value.getClass();

        if (type == String.class) { sb.append('"').append(escapeString((String) value)).append('"'); return; }
        if (type == Integer.class || type == Long.class) { sb.append(value); return; }
        if (type == Double.class) { sb.append(String.format(Locale.ROOT, "%s", value)); return; }
        if (type == Float.class) { sb.append(String.format(Locale.ROOT, "%s", value)); return; }
        if (type == Boolean.class) { sb.append((Boolean) value ? "true" : "false"); return; }

        // List
        if (value instanceof List<?> list) {
            sb.append('[');
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) sb.append(", ");
                serializeObject(list.get(i), sb, pretty, indent);
            }
            sb.append(']');
            return;
        }

        // Map
        if (value instanceof Map<?, ?> map) {
            if (pretty) serializeMapPretty(map, sb, indent);
            else serializeMapCompact(map, sb);
            return;
        }

        // POJO
        var fields = type.getDeclaredFields();
        if (pretty) {
            var inner = " ".repeat(indent + 4);
            sb.append('\n');
            for (var field : fields) {
                if (Modifier.isStatic(field.getModifiers())) continue;
                field.setAccessible(true);
                try {
                    var val = field.get(value);
                    sb.append(inner).append(WclDeserializer.toSnakeCase(field.getName())).append(" = ");
                    serializeObject(val, sb, pretty, indent + 4);
                    sb.append('\n');
                } catch (IllegalAccessException e) {
                    // skip
                }
            }
        } else {
            boolean first = true;
            for (var field : fields) {
                if (Modifier.isStatic(field.getModifiers())) continue;
                field.setAccessible(true);
                try {
                    if (!first) sb.append('\n');
                    first = false;
                    var val = field.get(value);
                    sb.append(WclDeserializer.toSnakeCase(field.getName())).append(" = ");
                    serializeObject(val, sb, false, 0);
                } catch (IllegalAccessException e) {
                    // skip
                }
            }
        }
    }

    private static void serializeMapCompact(Map<?, ?> map, StringBuilder sb) {
        sb.append('{');
        boolean first = true;
        for (var entry : map.entrySet()) {
            if (!first) sb.append(", ");
            first = false;
            sb.append(entry.getKey()).append(" = ");
            serializeObject(entry.getValue(), sb, false, 0);
        }
        sb.append('}');
    }

    private static void serializeMapPretty(Map<?, ?> map, StringBuilder sb, int indent) {
        var inner = " ".repeat(indent + 4);
        sb.append("{\n");
        for (var entry : map.entrySet()) {
            sb.append(inner).append(entry.getKey()).append(" = ");
            serializeObject(entry.getValue(), sb, true, indent + 4);
            sb.append('\n');
        }
        sb.append(" ".repeat(indent)).append('}');
    }

    private static String escapeString(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"")
                .replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t");
    }
}
